xds/src/main/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManager.java

```diff
@@ -58,97 +59,76 @@
+      case PREFIX:
+        return verifyDnsNamePrefix(
+            altNameFromCert, sanToVerifyMatcher.getPrefix(), sanToVerifyMatcher.getIgnoreCase());
+      case SUFFIX:
+        return verifyDnsNameSuffix(
+            altNameFromCert, sanToVerifyMatcher.getSuffix(), sanToVerifyMatcher.getIgnoreCase());
+      case CONTAINS:
+        return verifyDnsNameContains(
+            altNameFromCert, sanToVerifyMatcher.getContains(), sanToVerifyMatcher.getIgnoreCase());
+      case SAFE_REGEX:
+        return verifyDnsNameSafeRegex(altNameFromCert, sanToVerifyMatcher.getSafeRegex());
+      default:
+        throw new IllegalArgumentException(
+            "Unknown match-pattern-case " + sanToVerifyMatcher.getMatchPatternCase());
     }
-    // Wildcard pattern
+  }
 
-    // WILDCARD PATTERN RULES:
-    // 1. Asterisk (*) is only permitted in the left-most domain name label and must be the
-    //    only character in that label (i.e., must match the whole left-most label).
-    //    For example, *.example.com is permitted, while *a.example.com, a*.example.com,
-    //    a*b.example.com, a.*.example.com are not permitted.
-    // 2. Asterisk (*) cannot match across domain name labels.
-    //    For example, *.example.com matches test.example.com but does not match
-    //    sub.test.example.com.
-    // 3. Wildcard patterns for single-label domain names are not permitted.
+  private static boolean verifyDnsNameSafeRegex(
+          String altNameFromCert, RegexMatcher sanToVerifySafeRegex) {
+    Pattern safeRegExMatch = Pattern.compile(sanToVerifySafeRegex.getRegex());
+    return safeRegExMatch.matches(altNameFromCert);
+  }
 
-    if (!pattern.startsWith("*.") || pattern.indexOf('*', 1) != -1) {
-      // Asterisk (*) is only permitted in the left-most domain name label and must be the only
-      // character in that label
+  private static boolean verifyDnsNamePrefix(
+      String altNameFromCert, String sanToVerifyPrefix, boolean ignoreCase) {
+    if (Strings.isNullOrEmpty(sanToVerifyPrefix)) {
       return false;
     }
+    return ignoreCase
+        ? altNameFromCert.toLowerCase().startsWith(sanToVerifyPrefix.toLowerCase())
+        : altNameFromCert.startsWith(sanToVerifyPrefix);
+  }
 
-    // Optimization: check whether sanToVerify is too short to match the pattern. sanToVerify must
-    // be at
-    // least as long as the pattern because asterisk must match the whole left-most label and
-    // sanToVerify starts with a non-empty label. Thus, asterisk has to match one or more
-    // characters.
-    if (sanToVerify.length() < pattern.length()) {
-      // sanToVerify too short to match the pattern.
+  private static boolean verifyDnsNameSuffix(
+          String altNameFromCert, String sanToVerifySuffix, boolean ignoreCase) {
+    if (Strings.isNullOrEmpty(sanToVerifySuffix)) {
       return false;
     }
+    return ignoreCase
+            ? altNameFromCert.toLowerCase().endsWith(sanToVerifySuffix.toLowerCase())
```

Comment:
Specify a Locale?

Reply:
Hmmm, so `equalsIgnoreCase` uses the default Locale but you think when we do just prefix/suffix matching we should use a fixed locale - like `Locale.ENGLISH`? I don't know what the right answer is since this can be seen either way and [this](https://github.com/envoyproxy/envoy/blob/master/api/envoy/type/matcher/v3/string.proto#L69) doesn't say anything. But we should be consistent everywhere

Reply:
My understanding is `toLowerCase()` uses `Locale.getDefault()`, the value of which JVM specific. Whereas `equalsIgnoreCase()` does not take `Locale` in to account, so for ascii chars this is equivalent to using `Locale.ENGLISH` or `Locale.US`.

Reply:
Looking at this spec for [equalsIgnoreCase](https://docs.oracle.com/javase/9/docs/api/java/lang/String.html#equalsIgnoreCase-java.lang.String-) : 

> Calling Character.toLowerCase(Character.toUpperCase(char)) on each character produces the same result

It does say "...does not take Locale into account..."  but calling toLowerCase or toUpperCase() without passing the Locale value is like using the default Locale and *not* a fixed locale. You can also check the source code of `String.equalsIgnoreCase`

Reply:
The static method [`Character.toLowerCase()`](https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#toLowerCase(char)) is also Locale-insensitive, as `String.equalsIgnoreCase()`, as apposed to `String.toLowerCase()` which is Locale-sensitive and may have different behavior for different JVM.